### ISSUE TITLE
Should workers support py-* events?

### ISSUE THREAD

**Issue by antocuni** [pyscript/pyscript#1763]
This is a sub-issue of #1762.
I expect some discussion here, so better to open a separate one.

Currently, we have 4 tests which works in the main thread but not in the worker:

https://github.com/pyscript/pyscript/blob/c6aaacdbf17eb8e6d58f6d6cd8ccb4d511d211f0/pyscript.core/tests/integration/test_02_display.py#L174-L189

https://github.com/pyscript/pyscript/blob/c6aaacdbf17eb8e6d58f6d6cd8ccb4d511d211f0/pyscript.core/tests/integration/test_02_display.py#L208-L222

https://github.com/pyscript/pyscript/blob/c6aaacdbf17eb8e6d58f6d6cd8ccb4d511d211f0/pyscript.core/tests/integration/test_01_basic.py#L113-L140

https://github.com/pyscript/pyscript/blob/c6aaacdbf17eb8e6d58f6d6cd8ccb4d511d211f0/pyscript.core/tests/integration/test_01_basic.py#L357-L372

They all share the same fundamental reason:
```html
<button py-click="myfunc">Click me</button>
<script type="py" worker>
    def myfunc(event):
        print("hello world!")
</script>
```

Should this work or not?
Let me try to play devil's advocate for both camps.

**Team "it should not work"**

One reason for saying "not" is that it's unclear what to do in case of multiple workers, or maybe main-thread+workers, e.g.:
```html
<button py-click="myfunc">Click me</button>
<script type="py" src="aaa.py"></script>
<script type="py" src="bbb.py" worker></script>
<script type="py" src="ccc.py" worker></script>
```

In a situation like the above, we have three different interpreters: one in the main thread and one for each of the two separate workers, and it's not obvious/clear/well defined where to search and find `myfunc`, so better to avoid it entirely.

**Team "it should work"**

In case there is only one interpreter running in a worker, it is obvious where to look.
Moreover, we make it easier to switch from "main thread" to "worker", and it makes the two cases behaving more similarly.

Also, I expect that in the majority of cases people will have a single interpreter running anyway (either main thread or worker), so this is a solution which "just works" for the common case.

In case of multiple interpreters, we should think of a more explicit syntax and/or require them to use `@when`.

Personally, I am in the "it should work camp" :)

/cc @WebReflection @ntoll @tedpatrick @JeffersGlass

**Comment by WebReflection** [comment]
I'm team **it should not work** for the following reason: we export a `@when` decorator that supposed to be superior to `py-xxx` events and it should be clear that workers operate in a different world, they are not 1:1 related to main, no matter how much magic we introduce (magic that in this specific proposal screams for troubles).

However, for the **it should work** team I'd like to list all issues that comes to mind just now, without much further thinking of even more issues to consider:

> In case there is only one interpreter running in a worker, it is obvious where to look.

not at all ... those events are nowhere leaked in the main thread, those are worker defined methods and I've no idea how to expose those functions automatically to the outer world out of user code. With main thread code those leak globally but the pyodide in a worker cannot be passed along a *postMessage* and it stays sandboxed in the worker ... we could orchestrate something like "*leak all global methods to the main world for listeners after a single worked is executed and map it by the interpeter or custom type name*" which *might* work but it requires a lot of checks and validations in the main thread around how many tags, what happens if a program add a tag at runtime, and so on ... this is the first slippery slope to me, but in an ideal static only HTML page this might be solved.

> we make it easier to switch from "main thread" to "worker", and it makes the two cases behaving more similarly.

from a user perspective, this is the only argument that makes me think that *maybe* we should at least try to investigate this possibility

> I expect that in the majority of cases people will have a single interpreter running anyway (either main thread or worker)

yes and no ... I expect `mpy` on the main thread any time and `py` in workers all other times when more complex stuff is needed ... luckily enough neither interfere with each other (`mpy-click` vs `py-click`) but it's worth mentioning this use case which I think will flourish in the near future.

> In case of multiple interpreters, we should think of a more explicit syntax

that's unnecessary ... if you have multiple custom types (`mpy` and `py`) nothing conflicts with each other ... so that if there is only *one* `<script type="py">` or one `<py-script>` on the page, nobody cares if `<script type="mpy">` or `<mpy-script>` tags are somewhere else ... the contract here is **1 source of truth only** and that's the per custom type where each custom type has only one interpreter still by contract so ... we're good there, this is not the issue we'll face to implement this.

> and/or require them to use `@when`.

I think `@when` should be suggested in general because it keeps logic and expectations within the *Python* world, separating concerns with the layout, but I also believe `py-click` and friends are too damn and deadly easy to sell so I understand where you come from but then again, there are technical difficulties to fix.

My mild push back is mostly to avoid bloating the current achievement of shipping a tiny library with stuff that's maybe both early and too new to decide we should spend time fixing issues we never had to date, but on the other hand, if you folks are happy with me taking a chance to implement this idea, I'd be game!

**Comment by WebReflection** [comment]
P.S. this issue feels like it should've been a discussion instead ... just saying.

**Comment by ntoll** [comment]
@antocuni this is a really important discussion to have, thank you for bringing it up.

I'm in the "it should NOT work" camp. I think there are simply too many edge cases or work-arounds and a feeling of "magic" (one abstraction too far IYSWIM) needed to make it work when, as @WebReflection suggests, there are already ways to achieve the same effect (use `@when` rather than `py-*`) and it would complicate the implementation.

Also, we can be very explicit in the documentation about HOW to handle such things in workers.

**Comment by WebReflection** [comment]
I'd like to add an extra limitation of the "too magic" world we'd like to offer:

  * there is no way to have 1:1 event listeners in *main* thread and *worker* thread, mostly due `event` lifecycle

To expand, the *main* thread cannot block the thread, it's not allowed by atomics, it's even disabled and deprecated by workarounds via synchronous XHR in the near future.

"*... and where is the issue?*" ... well, to start with, we cannot `event.preventDefault()` or `event.stopPropagation()` because by the time the worker receives the event from main this is kinda "*dead*" already. It will carry its own details, which is fine, but the borwser will throw an error or silently fail expectations because the event is forwarded in another world where it cannot prevent any default or stop any propagation ... so this is the biggest limitation out there that makes code impossible to guarantee as 1:1 from both worlds.

*coincident* solves this issue by providing [an extra field](https://github.com/WebReflection/coincident#extra-features) to the third argument of `addEventlistener` but that works procedurally, not imperatively, and by no mean those operations can be decided at runtime (for input cases, as example) as these must be known at event listener attach time, not a tick later.

We could find some hack/workaround to make then code able to understand those operations in both main and worker thread, but it must be clear the code itself cannot be 100% ported across these two worlds, and specially for events this limitation *must* be understood.

This "*gotcha*" is bigger than any "*how do I expose listeners to main*" friction because it's a limitation backed in by the very same standard Web specifications we use as platform, so if any other argument doesn't convince we shouldn't move forward with this idea, this technical limitation will be a nightmare to implement and explain to our users.

Again, I am the least person that would ever say "*it's impossible*", we live on the edge out of modern features and little hacks here and there, but the event system in the main thread is unfortunately what it is and I've asked already WHATWG to provide an `event.waitUntil(...)` utility while solving this in coincident and there were crickets: https://github.com/whatwg/html/issues/9540

**Comment by ntoll** [comment]
> I think we should really reconsider whether we want to have py-click at all: having two things which are similar-but-slightly-different sounds like a good recipe to cause confusion.

@antocuni I wondered about this too... but `py-click` makes it very easy for "first steps" examples (i.e. it's explicit in the HTML that such-and-such a button will be handled by a Python function called whatever).

**Comment by antocuni** [comment]
> @antocuni I wondered about this too... but py-click makes it very easy for "first steps" examples (i.e. it's explicit in the HTML that such-and-such a button will be handled by a Python function called whatever).

yes, I agree that it's super nice -- that's why I want it for workers as well :)

But only if the semantics of py-click and `@when` are equivalent: if not, we need to discuss. Let's first try to understand if there are differences and what they are.

**Comment by WebReflection** [comment]
> This is supposed to work, isn't it?

not quite ... 

```python
@when("click", selector="#btn")
def onclick(event):
    # this works in main AND fails in workers
    event.preventDefault()
```

but `@when` is a declarative approach *and* it can use the *coincident* patch to make thing working from a worker:

```python
@when("click", selector="#btn", invoke=["preventDefault"])
def onclick(event):
    # this works in both main AND workers
```

the details is that `@when` can then decide to add a listener that `event.preventDefault()` out of the box in main and, when we're running in a worker, just pass along that `invoke` to `addEventListener` as third argument.

This requires documentation around but there is no fix for the first example I've shown because it will throw or not do what users wanted to do *out of Web standard constraints* so there's nothing we can do about it.

>  I think we should really reconsider whether we want to have `py-click` at all

absoltely **not** because `worker` is a great use case for *pyodide* but a completely unnecessary use case for *micropython* or whatever PL we'll like to offer in the future.

In all *main* thread scenarios the `xxx-click` event is an absolute wonder and ease to explain and implement for anything that is not in worker ... it seems to me it's still hard to not think *pyodide* only with current offer ... we should try harder to think more broadly, imho.

**Comment by antocuni** [comment]
> absoltely not because worker is a great use case for pyodide but a completely unnecessary use case for micropython or whatever PL we'll like to offer in the future.

I disagree. MicroPython has the very same problems as pyodide w.r.t. blocking calls.

My not-so-secret idea is that a future version of pyscript should switch to worker-by-default, that's why I'd like the two cases to behave as similar as they can.
(but let's not discuss "worker-by-default" here, I know it's a can of worms)

**Comment by WebReflection** [comment]
> MicroPython has the very same problems as pyodide w.r.t. blocking calls.

and so does JS ... that's why we can have asynchronous listeners ... MicroPython doesn't need worker because it doesn't really block in a relevant way the main thread on bootstrap. Performance on clicks are always good enough, being either pyodide or micropython, but pyodide can do much more thanks to `micropip`.

Asynchronous events still cannot `preventDefault` or `stopPropagation` at distance, these operations must be part of the still synchronous part and we **cannot** offer that due Web standard constraints.

**Comment by antocuni** [comment]
> MicroPython doesn't need worker because it doesn't really block in a relevant way the main thread on bootstrap.

well, it depends what you do. If you have an infinite loop, it surely blocks :)

But it's not only about the performance: we want `input()` in micropython as well.

I'm not saying that "micropython in the main thread" is not a valid use case, it totally is. But there are tons of reasons for wanting "micropython in the worker" as well

**Comment by antocuni** [comment]
> same in JS ... if you have an infinite loop there's no worker that can save your broken site or code, imho.

yes, because if the infinite loop is in a worker, we can have a button to kill/stop it, and possibly start a debugger.
We are targeting the famous 99%, and I'm pretty sure they will write infinite loops eventually :)

**Comment by ntoll** [comment]
I'm 100% sure the 99% will `while True:`.

**Comment by WebReflection** [comment]
To clarify, our users can write infinite loops all they like/want/need to understand that's an issue, but none of this helps solving the fact we cannot `preventDefault` or `stopPropagation` from a worker unless we take over the whole Event loop dance, which is not an easy (or imho desirable) task, and it has its own issues because of trusted events. https://developer.mozilla.org/en-US/docs/Web/API/Event/isTrusted

**Comment by antocuni** [comment]
I agree, the `while True` conversation is only tangential to this issue: it was brought in

**Comment by WebReflection** [comment]
> I think that the conclusion is that workers are a perfectly valid use case also for micropython.

your conclusion, maybe, mine differs ... I can bootstrap *mycropython* on main, write minimal DOM related code in it, offer all `mpy-click` and other events I want, and let *pyodide*, if necessary, do more heavy stuff, where such stuff can also trigger events on the main so that *micropython* has no issues whatsoever with the event native lifecycle.

Accordingly, I am for point 1: `leave everything as is` *except* the `@when` story should be improved, so that it can procedurally pass along `preventDefault` or other actions, when desired, and guarantee these works seamlessly from workers too.

On the main, `py-*` attributes live and exist on main only, so pretending these can exist in any form in a non-existent HTML or DOM on the Worker side, makes really zero to little sense to me ... as main is where they live, where these are defined, where these operate with their own thread.

`@when` is the feature that should cross boundaries, not attributes on the main HTML page, imho.

**Comment by ntoll** [comment]
FWIW :+1: for workers as a valid use case for MicroPython. Our users will surprise us. :wink:

**Comment by WebReflection** [comment]
then `@when` is what they should use, not `mpy-click` in the HTML. I don't think, from a Web standard point of view, I would ever change my mind about this and for backward compatibility that uses to work on main only (or feature parity) I don't think we should break that. There's no elegant solution neither to me but if anyone has one, let's discuss that, otherwise it's a technical Web standard limitation and those events to date worked fine for PyScript users.

**Comment by WebReflection** [comment]
as reminder, if [this was accepted](https://github.com/whatwg/html/issues/9540) this discussion wouldn't exist ... as long as that's ignored this discussion has only two way forwards:

  1. leave everything as is
  2. kill py-click

there's no middle ground for events here, unless you want me to spend weeks taking over the whole events native life-cycle and still be incapable of providing the `isTrusted` native meaning + it will slow everything down by default out of a polyfill for a standard that got rejected already ... it's your call if you want to consider the middle point @antocuni mentioned, otherwise we'll go for the super awkward and ugly `data-prevent-default` or `data-stop-propagation` and `data-stop-immediate-propagation` which is a slippery slope and future hostile to me for the project.

**Comment by JeffersGlass** [comment]
I think we can see (thank you @WebReflection!) that there's no way to have 1:1 event listeners in main thread and worker thread, for a number of important technical reasons (no `preventDefault`/`stopPropagation`, no way to emulate `isTrusted`, +others). Those are surely limitations, and worth documenting, but for me personally, they're not dealbreakers for getting py-[event] to work with workers. 

That is, saying: *"py-[event] also works with workers, but the following list of things don't work; if you need them, use `@when`* I think is reasonable...

As you said earlier, the bigger question is how do expose the names of those event handlers from the interpreters in workers, or how to establish 'which' worker to look up the event handler name in (or search "all the extant workers")... lots of ways of asking the same messy question.

You mention above of "export all the global names from the interpreter after execution" would be one way; another that had been tossed around long ago was:

```html
<!-- py-[event] syntax uses the main-thread interpreter -->
<script type="py">
    def normal(*args):
        print("A normal py-[event] handler")
</script>
<button py-click="normal">Click Me!</button>

<!-- To use attributes with worker threads (1) the worker must be given a name and (2) the attribute is py-[event]-[worker name] -->
<script type="py" worker worker-name="foo">
    def handler(*args):
        print("I am being handled!")
</script>
<button py-click-foo="handler">Click Me!</button>
```
Not saying this is good either 😓 it adds yet more attributes, they might be confusing, lots of edge cases as @ntoll says (how to we accomodate workers created by calling PyWorker directly? etc), but I'm wondering if some API design can help us solve the name-lookup problem.

**Comment by WebReflection** [comment]
I don't want to sound rude or anything but I feel like we really love to find ourselves in `fetch` like places where everything is not easy to explain, the branching logic is all over the place, and what users wanted were just events.

```html
<script type="py" worker worker-name="foo">
    def handler(*args):
        print("I am being handled!")
</script>
<button py-click-foo="handler">Click Me!</button>
```

this is asymmetric in all possible ways, it's not easy to explain, it adds complexity to users expectations compared to a "*use `@when` instead*" message for main thread attributes events, so that:

  * we need to add **a lot** of logic to solve the fact main knows nothing about workers globals and vice-versa
  * we add 4 half-working ways to solve the same topic instead of saying: `@when` is the best way we can offer, the rest is *main* only
  * the confusion around all custom attributes no Web standard document just increases exponentially
  * nobody is happy as we don't offer a solution and it's not simple to understand for users in terms of caveats and possibilities

> for me personally, they're not dealbreakers for getting py-[event] to work with workers.

neither for me, but I can see tons of issues filed because we couldn't `stopPropagation` or `preventDefault` out of forms submits because a user just added `worker` to their tag ... from a user perspective, I'd rather prefer that listener to never trigger than let it trigger at distance when the page is loading or submitting something else, as example ... so once again, it's your call!

More than exposing all the reasons this has never been an issue in the first place but this also can never works seamlessly in workers I am not sure how else I can contribute here ... but it's sure I won't be able to fix the un-fixable so whatever decision we take I'd love to have all expectations and failing points listed in that decision, thank you!

**Comment by WebReflection** [comment]
to play slightly more friendly, assuming we really want to go down this path, to which I still disagree, here my counter proposal:

```html
<script type="py" worker>
    def handler(*args):
        print("I am being handled!")
</script>
<button py-click="handler" worker>Click Me!</button>
```

  * the intent is explicit without messing around with extra attributes, as we already kinda decided the Web won't ever expose a `worker` attribute because if they do we might have issues
  * the most common use case (one script/tag as worker) out there is covered ... to me is really hard to think about any other "multiple workers" use cases, and I am sure if there is such use case that's not for the HTML, rather for imperative approaches ... *but* ...
  * keeping the same pattern, we can disambiguate with standards practices ... such as `id`

```html
<script type="py" id="foo" worker>
    def handler(*args):
        print("I am being handled!")
</script>
<button py-click="handler" worker="foo">Click Me!</button>
```

This is the only good-enough solution I could think about, because relations are all well defined, no new home-made attributes are needed, the relation between handlers and workers are crystal clear and last, but not least, there's less code to change, specially around the click handler type, as there's less guessing about the interpreter, as `py-click-foo` doesn't disambiguate if the interpreter name is `py-click` or something else, because we have `ruby-wasm-wasy` as handler too, as example, behind the scene, so that we disambiguate by exact worker target as id, as IDs can survive cross realm/threads dance.

If this proposal is OK to everyone, we still need to:

  * solve the issue that main thread knows nothing about workers globals (and vice-versa, but vice-versa is not interesting here)
  * change docs to explain worker listeners are inevitably asynchronous so that no `preventDefault` or `stopPropagation` are possible, among other native events properties or methods
  * be sure the default case VS the explicit ID case actually works and/or produce the right error when it doesn't

What do you all think, if you are still in the "*we should bring `py-*` to workers too**" ?

And a last question: are we going to make `@when` better at all? 'cause it still *could* offer even `preventDefault` or `stopPropagation` features when no condition is needed and that's desired out of the box, so I think we should consider improving that story too.

**Comment by ntoll** [comment]
Once again, a really stimulating read.

I'm going to put my flag in the `py-*` events should be killed off - only use `@when` in code so it's nice and explicit *where* the event handler is defined.

Why?

Two reasons... 

1. In the web world (please correct me if I'm wrong @WebReflection), folks advise NOT to use the `py-*` equivalent `onclick`, `ondblclick`, etc... HTML attribute means of defining an event handler. MDN is very clear, "[you should never use the HTML event handler attributes — those are outdated, and using them is bad practice](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#inline_event_handlers_%E2%80%94_dont_use_these)" and they only [describes the explicit definition of event handling in code](https://developer.mozilla.org/en-US/docs/Web/API/Element/click_event))
2. In the Python world, the Zen of Python is very clear:

> In the face of ambiguity, refuse the temptation to guess.
> There should be one-- and preferably only one --obvious way to do it.
> Although that way may not be obvious at first unless you're Dutch.

By killing these off we're being both idiomatic of the web AND Python, and a whole host of problems disappear.

Let's just get rid of them.

As always, I welcome constructive feedback! The technical discussions we have around PyScript are always a great opportunity to explore such interesting problems with formidably talented engineers. Bravo.

**Comment by WebReflection** [comment]
If we kill inline events there's a lot of code to remove in both *polyscript* and *PyScript* (specially integration tests but also example). I personally think our users will complain a lot about this decision because while it's true that inline HTML events are usually considered a bad practice and a code-smell, simply because events need to be leaked globally to be reached even via standard JS, it's a handy convention that doesn't require a unique ID for every single target (imagine a list of `<LI>` or `<TR>` with related actions, as example). That being said, making `@when` the way works but it's also a *PyScript* thing only so I think I should rather pass a flag to *polyscript* that eventually disables the ability to attach event to known interpreters and types or everything else that is not *PyScript* breaks: namely *wasmoon*, *ruby-wasm-wasi*, and *R* in the near future.

These languages can't even have decorators in the same *Python* meaning (AFAIK) and neither can JS (still, even if stage-something in the TC39 pipe). If it's not through our users we make this decision let's at least ensure some deprecation time and re-think the decision if everyone starts complaining about such deprecation ... shall we?

**Comment by WebReflection** [comment]
P.S. personally I think [this counter-proposal](https://github.com/pyscript/pyscript/issues/1763#issuecomment-1747160256) of mine would be the most desired and elegant way as long as we don't overengineer its functionality with improbable uses cases such as "*but how does anyone add a listener from one worker and another listener from another or another listener from main*" ... the logic is simple and it must be simple, while `@when` should still be the preferred way to go.

After all, the Web has inline events and these are there to stay. Everyone uses `addEventListener` these days as the recommended way but there are still early users just invoking global functions and if it's the 99% of people we're targeting, keeping it simple for them works too. Other libraries also just attach events directly as in `button.onclick = handler` so there are 3 ways to add listeners on the Web and *that is just fine* and all of them has *valid or simplified use cases* that *just work* too ... so I rather improve expectations that works in every programming language than kill the pattern that worked very well for our own examples, talks, demos and, most importantly, our users to date.

**Comment by WebReflection** [comment]
sorry for not packing all thoughts at once but these keep coming up ... even if we kill `py-*` events, the `@when` solution will not be able to be 1:1 with main use cases because of all the technical limitations I've already described ... so we kill something desired by our users and our own examples and demos or integration tests to still provide something that cannot possibly solve in a 1:1 the underlying issue around worker events. What are we really gaining in pursuing this pattern and to be actually honest, if there was a way already to provide events, what was the rationale to introduce `@when` then, as it created already 2 ways to do the same thing somehow? That rationale, and I am sure there is one, is the same reason I think both ways should work as that's been the case to date. End of my thoughts (until the next one :D)

**Comment by WebReflection** [comment]
if anyone is interested in how the Web currently works or what it offers here a summary https://gist.github.com/WebReflection/b404c36f46371e3b1173bf5492acc944

**Comment by JeffersGlass** [comment]
Two quick thoughts:

>  [inline event handlers are] a handy convention that doesn't require a unique ID for every single target (imagine a list of <LI> or <TR> with related actions, as example).

This I think is the primary reason to preserve py-events at least in some form. Additionally, I *think* it can makes handling simple event listeners in adding event handling when other reactive libraries are used to manipulate page elements (i.e. React, Vue etc)... but I'm sure @WebReflection and @bugzpodder can speak more directly, I'm out of my depth there.

------

I concur that @WebReflection's proposal is superior/cleaner than mine, and should supersede it. 👍 And 100% hear you, and agree, about avoiding overengineering it's functionality.

**Comment by WebReflection** [comment]
thanks @JeffersGlass for reading and listening, I really appreciate that :+1: 

if I can summarize my personal feelings:

  * I am still for keep everything as it is and explain that HTML inline listeners are for main only, use `@when` (and improve it) for the worker story ... but ...
  * if we decide to kill inline events instead, I'd rather advance my latest proposal to make these work for workers too and hear feedback in case the `preventDefault` or `stopImmediate/Propagation` story (or anything else sync events related) comes up (KISS first then iterate after feedback)

as summary, I know I backfired on @antocuni second point around "*change semantics*" but if we allow compromises, the only semantic to change is to enable `worker` attribute on already parsed and orchestrated nodes out there with the handy `py-*` feature we offer ... I think that was a good one, I'd be super sad in seeing this disappear, as I've heard the *99% users* story already a few times, and here if looks like we'd cut out 80% of current *PyScript* users instead, if we kill that feature.

**Comment by WebReflection** [comment]
last elephant in the room here, if worth anything at all talking about it, is that `@when` currently works on *pyodide* only, hwile my proposal will work out of *polyscript* so it will work for *MicroPython* interpreter too, among all others ... *polyscript* is indeed where this feature should land, abstracting away these kind of discussions in the future, when we'll eventually enable more interpreters too, such as *R* which indeed works only asynchronously at this point but it could be unlocked by my proposal.

**Comment by bugzpodder** [comment]
honestly the `py-*` and `@when` feels a bit magical to me (i still don't know how it really works) but I find it useful for beginner developers who wants to use pyscript to build an interactive app.

Slightly offtopic:  javascript developers may not want to to use pyscript  into an existing app since they'll usually want to handle dom events in the framework (rather than through py-dom or py-events), having myself spending quite a bit of time on this.  In particular, as an app developer I probably want the following feature set
- a pure js API (pyodide/polyscript)
- worker support (polyscript provides this)
- input (pyodide)
- ffi (pyodide, or polyscript if in worker)
- display/current_target/module bootstrap (pyscript)
since polyscript lets you access pyodide directly so anything available in pyodide is available through polyscript.
at the end of day I have settled on polyscript with a light display/current_target/module bootstrap wrapper ripped from pyscript.

**Comment by JeffersGlass** [comment]
I can get behind this answer @fpliger! To the question: “should py-*events support workers?” the answer of “not in this release” makes good sense! 

A very good discussion though, let’s bring it back up when we’ve got a little more clear time?

**Comment by ntoll** [comment]
As always, what a great discussion... and thank you all for your careful, thoughtful and crafted responses. :+1:

Yes - I agree @fpliger, now is not a good time to change the universe for all the reasons we state.

Yes - I agree @WebReflection, this sort of behaviour feels better in defined in PolyScript (and thus cross interpreter) rather than in PyScript.

Yes - I see why "for the 99%" is a seemingly good argument. I'd push back and say that while something may appear easy for beginners, if it ingrains a bad practice **that needs to be unlearned later**, we're not helping beginners. We're just making it more difficult down the line and we shouldn't do it (I'm putting my educator's hat on here - trust me, we don't want to do this). However, I also concede that I'm not sure if attribute based event definition is bad practice *in* PyScript. I'd like to explore this further, base this on real world behaviour observed in learning situations, and refine according to evidence. As Fabio said, now is not the time or way to make this decision - but it should be there to discuss as we develop PyScript further.

Yes - using a decorator is Python specific, but it's just an idiomatic Python pattern to hide the `foo.addEventListener` call used to link events and handlers. Within PyScript we should totally do this (it's Pythonic!), and by moving the common aspects of this to PolyScript we're enabling other interpreters the choice to do whatever is idiomatic for them.

Onwards. :rocket:

**Comment by WebReflection** [comment]
> by moving the common aspects of this to PolyScript

not sure I am following but `@when` is not a polyscript matter to me and polyscript will keep having events as attributes as that's the only common behavior it can guarantee, not decorators.

anyway, let's discuss this when the time is ready, but around this point in particular:

> I'm not sure if attribute based event definition is bad practice in PyScript

it's not, because we don't evaluate attributes like the Web does, but it's via attribute that we orchestrate everything in PyScript.

on top of that, as last argument in favor of current behavior, one could just fetch a file copied and pasted without needing to change it at all and use any exported method from that file just by pointing at its method/utility ... we lose this portability across projects if we enforce the usage of `@when` and once again, `@when` doesn't guarantee any better practice because in workers it will be as broken as my latest proposal when it comes to expired events.

**Comment by WebReflection** [comment]
@bugzpodder 

> at the end of day I have settled on polyscript with a light display/current_target/module bootstrap wrapper ripped from pyscript.

please note *polyscript* exports a [target](https://github.com/pyscript/polyscript/blob/main/esm/worker/_template.js#L110-L124) which is what `current_target()` returns in PyScript. PyScript ensures that it is an ID in either main or workers while `target` for main returns the reference to the object (and maybe I should instead align these in polyscript to have the same meaning for target and then update PyScript to return an ID as we pass IDs in our APIs) ... what do you think? Apologies for the off topic though, maybe I should open a discussion in polyscript instead.

**Comment by Neon22** [comment]
I am so late here :( but we really need some examples of what's actually been implemented...  please :)
Especially where `@when` is used with a UI that adds/removes items from the dom that would be affected by that `@when`.

**Comment by WebReflection** [comment]
@Neon22 the `when` works and it's not about main thread attributes, not sure I am following

**Comment by JeffersGlass** [comment]
@Neon22  `@when` does not work, and is not intended to work, with dynamic changes to the DOM. It runs once, hooks up event listeners to the DOM objects that are selected, and that's it. It's a fairly thin wrapper around `aroundEventListener`. That's all it's meant to do.

If you want some kind of Pythonic way to hook up event listeners in the DOM (and given how many times it's come up, I know you do 😄) please file a Feature Request discussion describing the functionality you want to see.

**Comment by ntoll** [comment]
@Neon22 just to be clear... you are in no way becoming too annoying. Your contributions are greatly valued and we really appreciate you explaining to us both what you want and how you expected what we currently have would work (if there's a difference between what we have and how you expected it to work, we've not explained ourselves very well so your insight if invaluable). Keep on doing this! :+1:

**Comment by WebReflection** [comment]
should this follow up here? https://github.com/pyscript/pyscript/discussions/1845

about `py-*` events, `@when` is unrelated to that + I think we decided to promote `@when` instead of inline listeners, because these can't really work with worker scripts and/or multiple interpreters unless shenanigans are to be expected.

**Comment by JeffersGlass** [comment]
@Neon22 What Nicholas say is absolutely true! Sorry if my cheekiness was rude - I know you've been working with PyScript and dynamic pages, and truly it's really been very helpful in understanding what the limitations of our current APIs are. I just wanted to clarify that the reason there aren't examples of using `@when` with dynamic UI's is that that's outside the scope of what `when` is meant to accomplish. So if we need/want a Pythonic way to handle dynamic event listeners, let's all cook one up together, instead of saying that `@when` is wrong/broken.

Thank you for opening the discussion!

**Comment by Neon22** [comment]
no no not rude at all. Lets just say that in my long experience (worked for Symbolics), becoming an annoyance to developers is a fine line between advocating for change and hammering that nail until its a smashed flattened dull dead thing of no use to anyone at all anymore. :) I appreciate you all doing this work. Its a total game changer - which some people don't realise yet.

**Comment by ntoll** [comment]
@Neon22 wow... Symbolics... so much creative computing and coding came out of there. I hope we can be equally creative and empowering with PyScript and everything else in its orbit and ecosystem. I totally hear you about the smashed flattened dull dead thing... but that you said it, and we recognise it, is also a good sign of healthy coder-y contextual awareness. ;-)

**Comment by WebReflection** [comment]
Just passing through as maintainer duty, I think this issue is hard to follow, hard to answer to specific points, hard to move forward without confusion, from either implementors or users expectations.

I hence believe this should rather be closed and, if ever opened again, we should have a discussion instead.

To help everyone with it, I have opened a discussion that cross references this issue, so at least previous time is preserved.

https://github.com/pyscript/pyscript/discussions/1934
